### ISSUE TITLE
Snapshots from different cameras overwriting each other

### ISSUE THREAD

**Issue by PossumTV** [Motion-Project/motion#102]
I currently have a system which handles seven cameras and I've recently upgraded to this fork of Motion. I have found that when snapshots on different cameras are triggered simultaneously via http requests, the snapshots sometimes get corrupted and sometimes a snapshot taken on one camera will appear on another.

After digging into the code, it appears the cause is that the function **myfopen()** in **motion.c** isn't reentrant, due to the use of the global **buffers[]** structure. If called reentrantly, this function can sometime allocate the same buffer to two different jpegs, thus causing one to partly or wholly overwrite the other.

As an experiment, I put a mutex around the section in myfopen() which accesses the buffers[] structure and this seemed to have cured the problem, however I'm not sure what the best solution is. Maybe the buffers[] structure could be local to each thread? 

Also, I'd be interested to know how much of a performance boost you actually get out of this setvbuf() business. Has this been quantified?

**Comment by tosiara** [comment]
Nice catch!
Corrupted image issue when using multiple cameras has been reported many times for years:
http://www.lavrsen.dk/foswiki/bin/view/Motion/SupportQuestion2015x01x20x164957
http://www.lavrsen.dk/foswiki/bin/view/Motion/SupportQuestion2012x02x27x155712
http://www.lavrsen.dk/foswiki/bin/view/Motion/BugReport2011x06x24x150605

I hope that we finally identified the root cause. Would like to hear comments from @Mr-Dave

**Comment by tosiara** [comment]
@PossumTV is that what you get?

![two images mixed](http://img-fotki.yandex.ru/get/6201/73339514.1c/0_6616a_7e0a4ea9_orig.jpg)

**Comment by PossumTV** [comment]
My images are very small (352 x 288) and mostly what I would see is one image completely overwriting another, although occasionally I'd get something like part of one image followed by part of another. In the latter case it would look fairly similar to what you've shown; there would be a distinct chunk of one image followed by some banding like you've got and then a chunk of another image. Unfortunately I didn't save any samples.

Also, FWIW I have used 3.2.11 for several years prior to this and had never witnessed this problem with that version. It was by comparing the code between the two versions that I found what I believe to be the cause.
